frontend/src/metabase/admin/permissions/permissions.js

```diff
@@ -90,36 +93,56 @@ export const loadCollectionPermissions = createThunkAction(
   },
 );
 
-export const LIMIT_DATABASE_PERMISSION =
-  "metabase/admin/permissions/LIMIT_DATABASE_PERMISSION";
-export const limitDatabasePermission = createThunkAction(
-  LIMIT_DATABASE_PERMISSION,
-  (groupId, entityId, accessPermissionValue) => dispatch => {
-    const newValue =
-      PLUGIN_ADVANCED_PERMISSIONS.getDatabaseLimitedAccessPermission(
-        accessPermissionValue,
-      );
-
-    if (newValue) {
-      dispatch(
-        updateDataPermission({
-          groupId,
-          permission: { type: "access", permission: "data" },
-          value: newValue,
-          entityId,
+export const GRANULATE_DATABASE_TABLE_PERMISSIONS =
+  "metabase/admin/permissions/GRANULATE_DATABASE_TABLE_PERMISSIONS";
+export const granulateDatabasePermissions = createThunkAction(
+  GRANULATE_DATABASE_TABLE_PERMISSIONS,
+  (groupId, entityId, permission, value, defaultValue) =>
+    async (dispatch, getState) => {
+      // NOTE: table data needs to be loaded in order for downstreams functions
+      // to have all tables ids to key permissions by their ids
+      await dispatch(
+        Tables.actions.fetchList({
+          dbId: entityId.databaseId,
+          include_hidden: true,
+          remove_inactive: true,
         }),
       );
-    }
 
-    dispatch(push(getGroupFocusPermissionsUrl(groupId, entityId)));
-  },
+      // HACK: updatePermission fn in metabase/admin/permissions/utils/graph sets child entities of
+      // entities becoming controlled to the current value of the entity. in some cases the current
+      // value is one we can't allow the child entity to become, so we quickly update the entity's
+      // value to the prefered default here, then proceed with the actual update
```

Comment:
I would take another look at this and maybe go into a bit more detail so it's very clear what is going here. I'm curious why there are certain disallowed values for the child entity. (In what sense *can't* we allow the child entity to become that value? What would happen if we tried? Would that cause a syntax error?)

"sets child entities of entities becoming controlled to the current value of the entity" The grammar here is complicated. Maybe split this into two sentences?

Reply:
I really appreciate this type of feedback!